proposals/0000-integrations.md

```diff
@@ -0,0 +1,171 @@
+
+Adding a first-class integration system would unlock a few huge wins for Astro:
+1. Empower users to do more by extending Astro themselves (not blocked by what Astro core can/can't do).
+2. Empower users to do more by reusing shared extensions (easy to add "X" to an Astro project).
+3. Empower more user-land experimentation, reducing how much Astro core blocks what a user can/can't do with Astro.
+4. Organize our codebase by refactoring more logic into internal integrations or moved out of core entirely into external integrations.
+
+To illustrate this, compare Partytown's documentation for getting started wtih Astro vs. getting started with Nuxt:
+- Astro: [3 steps across frontmatter, template, and a custom npm script](https://partytown.builder.io/astro)
+- Nuxt: [1 step](https://partytown.builder.io/nuxt)
+
+Tailwind suffers from a similar difficult setup story in Astro.
+
+# Detailed design
+
+**Background:** This API was reached through weeks of experimentation of different designs (see alternatives below). To test the work, I designed and build the following integrations, which are useful for illustrating this RFC: 
+
+- **Renderers:** [`lit`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/lit/index.js), [`svelte`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/svelte/index.js), [`react`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/react/index.js), [`preact`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/preact/index.js), [`vue`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/vue/index.js), [`solid`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/solid/index.js)
+- **Libraries:** [`tailwind`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/tailwind/index.js), [`partytown`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/partytown/index.js), [`turbolinks`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/turbolinks/index.js)
+- **Features:** [`sitemap`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/sitemap/index.js)
+
+## Integration Usage API
+
+```js
+// astro.config.js
+export default ({
+	integrations: [
+		import('@astrojs/vue'),
+        // or, with options
+		[import('@astrojs/vue'), {/* options */}],
+        // or, as a static ESM import at the top of the file
+		vuePlugin({/* options */})
+	],
+});
+```
+
+## Integration Author API
+
+```ts
+export interface AstroIntegration {
+	name: string;
+	hooks: {
+		/** SETUP */
+		/** Called on astro startup, lets you modify config */
+		'astro:config:setup': (options: ConfigSetupOptions) => void | Promise<void>;
```

Comment:
Can you add the interface `ConfigSetupOptions` so we can see what is possible in this hook?

Reply:
```ts
{
  /** not-yet-completed Astro config object, can be modified in-place */
  config: AstroConfig;
  /** The astro command that was run. */
  command: 'dev' | 'build';
  /** Add a renderer (ex: React, Svelte). */
  addRenderer: (renderer: AstroRenderer) => void;
  /** Inject a script. */
  injectScript: (stage: 'beforeHydration' | 'head' | 'bundle', content: string) => void;
  /** Inject an HTML element. */
  injectHtml: (stage: 'head' | 'body', element: string) => void;
}
```

I left this out because this is still being finalized and still subject to change, and hoped we could consider it an implementation detail of the larger integration system. I'm  open to all feedback on this now but would also love to keep working on it up the the final PR.